crmd/notify.c

```diff
@@ -20,154 +20,721 @@
+                               g_list_first(environment_variable_list));
+    }
+    return environment_variable_list;
+}
+
+static GListPtr
+copy_envvar_list_remove_dupes(GListPtr src)
+{
+    GListPtr dst = NULL, ls, ld;
+
+    /* we are adding to the front so variable dupes coming via
+     * recipient-section have got precedence over those in the
+     * global section - we don't expect that many variables here
+     * that it pays off to go for a hash-table to make dupe elimination
+     * more efficient - maybe later when we might decide to do more
+     * with the variables than cycling through them
+     */
+
+    for (ls = g_list_first(src); ls; ls = g_list_next(ls)) {
+        for (ld = g_list_first(dst); ld; ld = g_list_next(ld)) {
+            if (!strcmp(((environment_variable_t *)(ls->data))->name,
+                        ((environment_variable_t *)(ld->data))->name)) {
+                break;
+            }
+        }
+        if (!ld) {
+            dst = g_list_prepend(dst, 
+                    copy_environment_variable_entry(
+                    (environment_variable_t *)(ls->data), NULL));
+        }
+    }
+
+    return dst;
+}
+
+static void 
+add_dup_notify_list_entry(notify_entry_t *entry)
+{
+    notify_entry_t *new_entry = 
+        (notify_entry_t *) calloc(1, sizeof(notify_entry_t));
+
+    CRM_ASSERT(new_entry);
+    *new_entry = (notify_entry_t) {
+        .id = strdup(entry->id),
+        .path = strdup(entry->path),
+        .timeout = entry->timeout,
+        .tstamp_format = entry->tstamp_format?strdup(entry->tstamp_format):NULL,
+        .recipient = entry->recipient?strdup(entry->recipient):NULL,
+        .environment_variables = entry->environment_variables?
+            copy_envvar_list_remove_dupes(entry->environment_variables)
+            :NULL
+    };
+    notify_list = g_list_prepend(notify_list, new_entry);
+}
 
-    if(cvalue == NULL) {
+static GListPtr
+get_envvars_from_cib(xmlNode *basenode, GListPtr list, int *count)
+{
+    xmlNode *envvar;
+    xmlNode *pair;
+
+    if ((!basenode) || 
+        (!(envvar = first_named_child(basenode, XML_CIB_TAG_NOTIFY_ENVVAR)))) {
+        return list;
+    }
+
+    for (pair = first_named_child(envvar, XML_CIB_TAG_NVPAIR);
+         pair; pair = __xml_next(pair)) {
+
+        environment_variable_t envvar_entry = (environment_variable_t) {
+            .name = (char *) crm_element_value(pair, XML_NVPAIR_ATTR_NAME),
+            .value = (char *) crm_element_value(pair, XML_NVPAIR_ATTR_VALUE)
+        };
+        (*count)++;
+        list = add_dup_envvar(list, &envvar_entry);
+    }
+
+    return list;
+}
+
+void
+notifications_query_callback(xmlNode * msg, int call_id, int rc, 
+                             xmlNode * output, void *user_data)
+{
+    xmlNode *notify = output;
+    notify_entry_t entry;
+
+    free_notify_list();
+
+    if (rc != pcmk_ok) {
+        crm_notice("No optional alerts section in cib");
```

Comment:
I think we can lower this (and the corresponding log below) to info level. In general, notice is to help sysadmins troubleshoot, while info is to help us troubleshoot.